### ISSUE TITLE
allow for silent authentication after signup

### ISSUE THREAD

**Issue by coffeedoughnuts** [auth0/auth0-spa-js#939]
### Describe the problem you'd like to have solved

I am implementing a custom sign up screen in my app and after my backend has created the user I want to authenticate the user and redirect them to my app without them needing to enter their password immediately after signup

### Describe the ideal solution

- my backend api has a `POST /signup` route that I send my user details (including password) to. This creates the user, then gets the user's access token and id token and returns it to the SPA
- my frontend app takes these tokens and provides them to the auth0 SDK to both verify and store
- my frontend app can then redirect my user to the app's secure area and they will not be asked to log in

## Alternatives and current work-arounds

my best current workaround is to use a localstorage cache and inject the codes in a way that matches the SDK's key and value format. This is not very good because that format is not part of the public API of the SDK and is liable to change without warning. Because of this, I don't want to implement this workaround in production.

### Additional context

If there is a better supported method of achieving the outcome I'm happy to change any part of my plan - my only hard requirement is that the signup UI must live in my app (not on an Auth0 page) and the signup API must be in my backend because I need to perform non-auth0 tasks when creating a new user.

**Comment by frederikprijck** [comment]
Thanks for reaching out!

Our SDK is intended to be used to retrieve and manage the tokens for you. We don't offer a supported way to inject externally retrieved tokens into our SDK, as that would open for doing things like `password_grant` (or any other grant for that matter) with our SDK which is strongly discouraged in Single Page applications.

Having said that, and trying to understand the situation, what's the reason the signup page has to exist in your app?
We would typically recommend, when working with SPA's, to not track credentials in your application but pass that responsibility to the Identity Provider in question.

Regarding executing custom non-auth0 tasks after creating a user, when you would sign up using Auth0, you can leverage [Actions](https://auth0.com/docs/customize/actions) and execute code after every user registration event, which could then call into your own API.

![image](https://user-images.githubusercontent.com/2146903/182351461-388eee7e-5fac-4501-8fbb-667efcaa356a.png)

**Comment by coffeedoughnuts** [comment]
thanks for getting back to me @frederikprijck 

our SPA is a large NextJS app built and exported statically (no server side stuff) and then our backend is a Clojure GraphQL app. We currently don't have public signup (it's handled manually as contracts are signed) but are now implementing the ability for existing users to invite new users to their organisation (our concept of organisations, not the auth0 feature)

we want the sign up page to live in our SPA so we can reuse the components, styles and build pipeline we currently have. We use custom domains for the login screen currently, but its a separate subdomain from the app (the app is at hub.company.com and auth0 pages are at auth.polecat.com)

are you suggesting that we instead move the signup pages to an auth0 template file which is available at auth.company.com ? does the client available in those template pages allow for signup + auth in one step? if so, can you link to some docs for that?

It would be less hassle to keep our signup page in our app (it's a multi-screen signup process and a bunch of logic to validate the invite, rather than just 'choose a password') but I suppose I could do a second NextJS export of just those pages and upload them to Auth0 on each build. Less keen on all our backend logic happening via an Action as it's not very transactional. We'd need to handle users existing in Auth0 but the corresponding data not existing elsewhere I suppose.

also, is it that rare a use case that we would want to own the sign up UX and have a smooth transition from sign up to logged in? I'm sort of surprised this isn't more of a supported feature, but I understand I could have overlooked an aspect of the security of the auth pipeline here.

**Comment by frederikprijck** [comment]
I believe there are two different things being discussed here.

**Injecting a token:**
We don't allow injecting a token obtained externally because of the fact that it opens the SDK to be (mis)used with flows that are not encouraged for Single Page Applications.
[As mentioned in the readme](https://github.com/auth0/auth0-spa-js), this SDK uses the Authorization Code flow using PKCE, retrieving a token is done using two steps:
- Retrieve an Authorization Code
- Exchange the Authorization Code for a Token

Apart from that, [security best practices]([security best practises draft](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-browser-based-apps#section-5)) also tells us we should retrieve a token in the two-step process as we are doing in our SDK. Returning such token from some other endpoint, that happens to do signup and lives outside of the Identity Provider, doesn't seem to fall into that recommendation.

Even more so, depending on your configuration, the SDK might rely on an active session with Auth0. Not sending the user to Auth0 might mean they have no Auth0 session, resulting in being logged out after the access token expires (which should be a matter of minutes).

Ofcourse, this can be mitigated when u use refresh tokens, but wanted to be sure to mention that as well.

**Custom signup page:**
Our SDK doesn't offer anything for that, and I am not specifically recommending any approach, I am mostly trying to understand your situation. We do have a professional services team who can work with you to understand the possibilities regarding custom signup flows.

What I do know is that we have the possibility to implement a custom signup after the initial signup at Auth0, using Actions. You can look at https://auth0.com/docs/customize/actions/flows-and-triggers/login-flow/redirect-with-actions

Basically you can hook into the pipeline, redirect to your page to obtain information beyond username and password, fill in those details, submit them, redirect back to Auth0, verify the data is available and continue the pipeline.

But again, this doesn't seem something we can action from the SDK side as we have no intention to allow injecting an access token because of the mentioned issue with opening for using unsecure flows.

**Comment by coffeedoughnuts** [comment]
thanks for the explanation - it makes sense that this SDK would not support the feature I've requested, so in that respect this issue can be closed.

I think our product team would need to talk to the professional services team to work out our best option. thanks for your help this all this!
